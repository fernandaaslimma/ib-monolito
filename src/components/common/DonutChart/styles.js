import styled, { css } from "styled-components";
import { rem, remFontSize, media } from "../../../styles/tools";
import { black50, white, blue30, gray300 } from "../../../styles/settings";

export const Wrapper = styled.div`
  border-radius: ${rem(4)};
  background-color: ${white};
`;

export const Totals = styled.div`
  padding-top: ${rem(5)};
  padding-right: ${rem(20)};
  padding-bottom: ${rem(5)};
  padding-left: ${rem(10)};
  margin-top: ${rem(5)};
  border-left: ${rem(1)} solid ${blue30};

  ${media.lg(css`
    padding-right: ${rem(33)};
  `)};
`;

export const ChartWrapper = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: ${rem(5)};
`;

export const ColoredDot = styled.span`
  width: ${rem(10)};
  height: ${rem(10)};
  border-radius: ${rem(3)};
  border: solid ${rem(1)} ${white};
  margin: auto ${rem(10)} auto 0;

  ${({ color }) =>
    color &&
    css`
      background-color: ${color};
    `};
`;

export const Item = styled.div`
  font-family: "Lato";
  color: ${black50};
  width: 100%;
  margin-bottom: ${rem(10)};
`;

export const Label = styled.div`
  font-family: "Lato";
  font-size: ${remFontSize(10)};
  text-transform: uppercase;
  display: flex;
`;

export const Value = styled.div`
  font-size: ${remFontSize(13)};
  margin-top: ${rem(3)};
  margin-left: ${rem(20)};
  letter-spacing: ${rem(0.4)};
`;

export const ChartTitle = styled.span`
  font-family: Lato;

  font-style: normal;
  font-weight: 600;
  font-size: ${remFontSize(14)};
  line-height: 125%;
  display: flex;
  justify-content: center;

  letter-spacing: ${rem(0.452308)};

  margin-top: ${rem(16)};
  margin-bottom: ${rem(16)};

  color: ${gray300};
`;
