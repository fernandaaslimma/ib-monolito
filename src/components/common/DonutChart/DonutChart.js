import "react-vis/dist/style.css";

import React, { useState, Fragment } from "react";
import { RadialChart, makeVisFlexible, Hint } from "react-vis";

import Card from "../../common/Card";
import { Wrapper, ChartWrapper, ChartTitle } from "./styles";
import formatNumber from "../../../utils/formatNumber";
import getColorForAsset from "../../../utils/getColorForAsset";
import { white } from "../../../styles/settings";
import ShimmerLoading from "../ShimmerLoading";

export const parseData = (assets) =>
  assets.map((asset, i) => ({
    angle: asset.grossBalance,
    label: `${asset.portfolioShare
        ? formatNumber(asset.portfolioShare)
        : formatNumber("0.00")
      } %`,
    portfolioShare: asset.portfolioShare,
    color: getColorForAsset(i)
  }));

function DonutChart({ assets, dataTest, children, title, loading }) {

  const FlexibleRadialChart = makeVisFlexible(RadialChart);

  const [value, setValue] = useState(null);
 
  const onValueMouseOver = (v) => setValue(v);
  const onValueMouseOut = () => setValue(null)

  return (
    <div data-test={dataTest}>
      <Wrapper>
        <Fragment>
          {title && (
            <ChartTitle data-test="typeGraphic">
              {loading ? <ShimmerLoading width={40} height={12} /> : title}
            </ChartTitle>
          )}

          <Card>
            {loading ? (
              <ShimmerLoading circle height={230} width={230} />
            ) : (
              <ChartWrapper onMouseLeave={() => setValue(null)}>
                {children}
                <FlexibleRadialChart
                  stroke={white}
                  colorType="literal"
                  height={230}
                  innerRadius={100}
                  radius={112}
                  data={parseData(assets)}
                  onValueMouseOver={onValueMouseOver}
                  onValueMouseOut={onValueMouseOut}
                >
                  {value && (
                    <Hint value={value}>
                    <div style={{ background: 'white', color: 'black', padding: 5, borderRadius: 5, fontFamily: 'Roboto', fontSize: 12 }}>
                      {value.label}
                    </div>
                    </Hint>
                  )}
                </FlexibleRadialChart>
              </ChartWrapper>
            )}
          </Card>
        </Fragment>
      </Wrapper>
    </div>
  );
}

export default DonutChart;
