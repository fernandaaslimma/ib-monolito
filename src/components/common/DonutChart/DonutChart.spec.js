import React from "react";
import { shallow } from "enzyme";

import DonutChart, { parseData } from "./DonutChart";

const consolidatedAssets = [
  {
    name: "Fixed Income",
    netBalance: 50,
    grossBalance: 50,
    portfolioShare: 80
  },
  {
    name: "Equities",
    netBalance: 50,
    grossBalance: 50,
    portfolioShare: 20
  },
  {
    name: "Cash Account",
    netBalance: 50,
    grossBalance: 50
  }
];

describe("AssetsChart mobile component", () => {
  it("should match snapshot loaded", () => {
    expect(
      shallow(
        <DonutChart
          assets={consolidatedAssets}
          title="Title"
          dataTest="Test"
          loading={false}
        />
      )
    ).toMatchSnapshot();
  });

  it("should match snapshot loading", () => {
    expect(
      shallow(
        <DonutChart
          assets={consolidatedAssets}
          title="Title"
          dataTest="Test"
          loading={true}
        />
      )
    ).toMatchSnapshot();
  });
});

describe("parseData", () => {
  it("should parse the data", () => {
    const assets = [
      {
        grossBalance: 123,
        portfolioShare: 15
      },
      {
        grossBalance: 2,
        portfolioShare: 50
      }
    ];
    expect(parseData(assets)).toEqual([
      { angle: 123, color: "#5b4a8a", label: "15 %", portfolioShare: 15 },
      { angle: 2, color: "#af5761", label: "50 %", portfolioShare: 50 }
    ]);
  });
});
 
