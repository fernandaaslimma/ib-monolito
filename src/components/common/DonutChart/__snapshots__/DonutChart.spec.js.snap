// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetsChart mobile component should match snapshot loaded 1`] = `
<div
  data-test="Test"
>
  <styled.div>
    <styled.span
      data-test="typeGraphic"
    >
      Title
    </styled.span>
    <Card
      icon={null}
      iconColor={null}
      title={null}
      titleColor={null}
    >
      <styled.div
        onMouseLeave={[Function]}
      >
        <FlexibleRadialChart
          colorType="literal"
          data={
            Array [
              Object {
                "angle": 50,
                "color": "#5b4a8a",
                "label": "80 %",
                "portfolioShare": 80,
              },
              Object {
                "angle": 50,
                "color": "#af5761",
                "label": "20 %",
                "portfolioShare": 20,
              },
              Object {
                "angle": 50,
                "color": "#96b974",
                "label": "0 %",
                "portfolioShare": undefined,
              },
            ]
          }
          height={230}
          innerRadius={100}
          onValueMouseOut={[Function]}
          onValueMouseOver={[Function]}
          radius={112}
          stroke="#ffffff"
        />
      </styled.div>
    </Card>
  </styled.div>
</div>
`;

exports[`AssetsChart mobile component should match snapshot loading 1`] = `
<div
  data-test="Test"
>
  <styled.div>
    <styled.span
      data-test="typeGraphic"
    >
      <ShimmerLoading
        darker={false}
        height={12}
        index={0}
        inverse={false}
        width={40}
      />
    </styled.span>
    <Card
      icon={null}
      iconColor={null}
      title={null}
      titleColor={null}
    >
      <ShimmerLoading
        circle={true}
        darker={false}
        height={230}
        index={0}
        inverse={false}
        width={230}
      />
    </Card>
  </styled.div>
</div>
`;
